codex/utils/asynciter.nim

```diff
@@ -49,98 +56,71 @@ proc new*[T](_: type Iter, genNext: GenNext[T], isFinished: IsFinished, finishOn
         iter.finish
       return item
     else:
-      raise newException(CatchableError, "Iterator is finished but next item was requested")
+      raise newException(CatchableError, "AsyncIter is finished but next item was requested")
 
   if isFinished():
     iter.finish
 
   iter.next = next
   return iter
 
-proc fromItems*[T](_: type Iter, items: seq[T]): Iter[T] =
-  ## Create new iterator from items
-  ##
-
-  Iter.fromSlice(0..<items.len)
-    .map((i: int) => items[i])
-
-proc fromSlice*[U, V: Ordinal](_: type Iter, slice: HSlice[U, V]): Iter[U] =
-  ## Creates new iterator from slice
-  ##
-
-  Iter.fromRange(slice.a.int, slice.b.int, 1)
-
-proc fromRange*[U, V, S: Ordinal](_: type Iter, a: U, b: V, step: S = 1): Iter[U] =
-  ## Creates new iterator in range a..b with specified step (default 1)
+proc emptyAsyncIter*[T](): AsyncIter[T] =
+  ## Creates an empty AsyncIter
   ##
 
-  var i = a
+  proc genNext(): Future[T] {.raises: [CatchableError].} =
+    raise newException(CatchableError, "Next item requested from an empty AsyncIter")
+  proc isFinished(): bool = true
 
-  proc genNext(): U =
-    let u = i
-    inc(i, step)
-    u
+  newAsyncIter[T](genNext, isFinished)
 
-  proc isFinished(): bool =
-    (step > 0 and i > b) or
-      (step < 0 and i < b)
-
-  Iter.new(genNext, isFinished)
-
-proc map*[T, U](iter: Iter[T], fn: Function[T, U]): Iter[U] =
-  Iter.new(
-    genNext    = () => fn(iter.next()),
+proc map*[T, U](iter: AsyncIter[T], fn: Function[T, Future[U]]): AsyncIter[U] =
+  newAsyncIter[U](
+    genNext    = () => iter.next().flatMap(fn),
     isFinished = () => iter.finished
   )
 
-proc filter*[T](iter: Iter[T], predicate: Function[T, bool]): Iter[T] =
-  var nextT: Option[T]
+proc mapFilter*[T, U](iter: AsyncIter[T], mapPredicate: Function[T, Future[Option[U]]]): Future[AsyncIter[U]] {.async.} =
```

Comment:
I guess I was expecting a map and a filter async equivalents as you had before. Why do we have a combined version as well, which appears to have slightly different semantics than combining map and filter (e.g. wrt error handling)?

Reply:
It should have exactly the same semantics. I can write a test for it if you want. The combined version is provided because it's a convenient API.

Reply:
Only if you feel you need to. I otherwise believe you.